src/dagre-js/create-nodes.js

```diff
@@ -46,10 +46,10 @@ var createNodes = function (selection, g, shapes) {
       labelGroup.attr('id', node.labelId);
     }
 
-    if (_.has(node, 'width')) {
+    if (Object.prototype.hasOwnProperty.call(node, 'width')) {
```

Comment:
Object.hasOwn is the replacement for this.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn

Reply:
Thanks for your feedback! @aloisklink was saying in a earlier thread in this page that he is not for it.

Reply:
I can go both directions. I'm more for `Object.hasOwn` to be honest.

Reply:
`Object.hasOwn` was only recently added to JavaScript (after the `7.0.0` release of `dagre-d3-es`).

My gut feeling is that if we do want to add it, we'd probably have to make a major version bump and release `dagre-d3-es` v8.0.0, just in case anybody is still using Node.JS v16 or an older browser.

Maybe I'm just being overly paranoid though :shrug:, but it's not too much extra effort to just find-and-replace all the `Object.hasOwn` calls with `Object.prototype.hasOwnProperty.call`.

Reply:
Speaking of which, it might be worth defining what JavaScript version this library supports.

The `tsconfig.json` says ES2020, https://github.com/tbo47/dagre-es/blob/805396cc76d1c81a43207ff9dca78945ccbf40f2/tsconfig.json#L6

If this library is aiming to be compatible with D3 v7, that still supports Node.JS v12: https://github.com/d3/d3/blob/4c01ba47177e3eb33066f87c835b9e161af1af54/package.json#L90 (although, that did come out in 2019, so it's not too long before ES2020)

Reply:
tsconfig.json and typescript are not used in the project yet.

I added it to be able to generate `*.d.ts`. But we are not using it yet.

The project is still a pure javascript project.

In the future, we could generate types from the source code by adding jsdoc for example. But we will probably want to be compatible with what there is in:
https://www.npmjs.com/package/@types/dagre
https://www.npmjs.com/package/@types/dagre-d3
https://www.npmjs.com/package/@types/graphlib

Reply:
> I added it to be able to generate *.d.ts. But we are not using it yet.

We are publishing the auto-generated types from the `.js` files (although very very very basic ones).

I had a look at https://arethetypeswrong.github.io/, and every single version of `dagre-d3-es` since version 5.0.1 seems to have types.

![image](https://github.com/user-attachments/assets/8e8dc2aa-c6a6-4ca5-a219-4ef270fbd8f6)

If we remove them, that would cause side-effects for people.

I've made a PR to re-add `*.d.ts` files in https://github.com/tbo47/dagre-es/pull/38! Thanks @tbo47 for reviewing it!

I've also checked and the `ES2020` does throw an error if we try to use an JavaScript features that were added more recently then ES2020!